cohttp-lwt/src/server.ml

```diff
@@ -111,43 +111,66 @@ module Make (IO : S.IO) = struct
     | Error e ->
         Log.info (fun m -> m "IO error while writing body: %a" IO.pp_error e);
         conn_closed ();
         Body.drain_body body
 
-  let rec handle_client ic oc conn spec =
-    Request.read ic >>= function
-    | `Eof ->
-        spec.conn_closed conn;
-        Lwt.return_unit
-    | `Invalid data ->
-        Log.err (fun m -> m "invalid input %s while handling client" data);
-        spec.conn_closed conn;
-        Lwt.return_unit
-    | `Ok req -> (
-        let body = read_body ic req in
-        handle_request spec.callback conn req body >>= function
-        | `Response (res, body) ->
-            let keep_alive = Request.is_keep_alive req in
-            handle_response ~keep_alive oc res body
-              (fun () -> spec.conn_closed conn)
-              (fun oc -> handle_client ic oc conn spec)
-        | `Expert (res, io_handler) ->
-            Response.write_header res oc >>= fun () ->
-            io_handler ic oc >>= fun () -> handle_client ic oc conn spec)
+  let handle_client ic oc conn spec =
+    let read_request () =
+      Request.read ic >>= function
+      | `Eof -> Lwt.return (None, None, false)
+      | `Invalid data ->
+          Log.err (fun m -> m "invalid input %s while handling client" data);
+          Lwt.return (None, None, false)
+      | `Ok req ->
+          let body = read_body ic req in
+          Lwt.return (Some body, Some req, true)
+    in
+    let handler request =
+      match request with
+      | None ->
+          let rec sleep_till_cancelled () =
+            Lwt_unix.sleep 1. >>= fun () -> sleep_till_cancelled ()
+          in
+          sleep_till_cancelled ()
+      | Some (body, req) -> (
+          handle_request spec.callback conn req body >>= function
+          | `Response (res, body) ->
+              handle_response oc res body (fun () -> spec.conn_closed conn)
+              >>= fun () -> Lwt.return (None, Some req, false)
+          | `Expert (res, io_handler) ->
+              Response.write_header res oc >>= fun () ->
+              io_handler ic oc >>= fun () -> Lwt.return (None, Some req, true))
+    in
+    (* Keep reading from the input channel while the client is processing the request.
+       If the EOF comes before the client finishes the processing, we will close the connection
+       and notify the client so that it can clean up allocated resources.
+       At first, we have no request to handle so we call handler() with None to keep it sleeping. *)
+    let rec handle_request_loop request =
+      (* read_request() can return None, None which means that the connection is closed and we need to stop.
+         If we got a new request and its body, we stop handling of the old request (if there was any) and
+         start handling of the new request.
+         If we got the requst only, it means that the handler has finished the handling. *)
+      Lwt.pick [ read_request (); handler request ] >>= fun handle_result ->
+      match handle_result with
```

Comment:
function? it will match the result without a variable, right?

Reply:
Updated, with another occurrence.